src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/BucketService.scala

```diff
@@ -0,0 +1,117 @@
+  def createInitBucket(googleProject: GoogleProject, clusterName: ClusterName, serviceAccountInfo: ServiceAccountInfo): Future[GcsBucketName] = {
+    val bucketName = generateUniqueBucketName(clusterName.value+"-init")
+    for {
+      // The init bucket is created in Leo's project, not the cluster's project.
+      _ <- googleStorageDAO.createBucket(dataprocConfig.leoGoogleProject, bucketName)
+
+      // Leo service account -> Owner
+      // User pet service account (notebook SA or cluster SA) -> Reader
+      petEntityOpt <- getPetOrDefault(googleProject, serviceAccountInfo)
+      leoEntity = userEntity(serviceAccountProvider.getLeoServiceAccountAndKey._1)
+
+      _ <- setBucketAcls(bucketName, petEntityOpt.toList, List(leoEntity))
+    } yield bucketName
+  }
+
+  /**
+    * Creates the dataproc staging bucket and sets the necessary ACLs.
+    */
+  def createStagingBucket(userEmail: WorkbenchEmail, googleProject: GoogleProject, clusterName: ClusterName, serviceAccountInfo: ServiceAccountInfo): Future[GcsBucketName] = {
+    val bucketName = generateUniqueBucketName(clusterName.value+"-staging")
+    for {
+      // The staging bucket is created in the cluster's project.
+      _ <- googleStorageDAO.createBucket(googleProject, bucketName)
+
+      // Leo service account -> Owner
+      // User pet service account (notebook SA or cluster SA) -> Owner
+      // Additional readers (users and groups) are specified by the service account provider.
+      leoEntity = userEntity(serviceAccountProvider.getLeoServiceAccountAndKey._1)
+      petEntityOpt <- getPetOrDefault(googleProject, serviceAccountInfo)
+      providerReaders <- serviceAccountProvider.listUsersStagingBucketReaders(userEmail).map(_.map(userEntity))
+      providerGroups <- serviceAccountProvider.listGroupsStagingBucketReaders(userEmail).map(_.map(groupEntity))
+
+      _ <- setBucketAcls(bucketName, providerReaders ++ providerGroups, List(leoEntity) ++ petEntityOpt.toList)
+    } yield bucketName
+  }
+
+  /**
+    * Sets ACLs on an existing user bucket so that it can be accessed in a dataproc cluster.
+    */
+  def updateUserBucket(bucketName: GcsBucketName, googleProject: GoogleProject, serviceAccountInfo: ServiceAccountInfo): Future[Unit] = {
+    for {
+      // User pet service account (notebook SA or cluster SA) -> Reader
+      petEntityOpt <- getPetOrDefault(googleProject, serviceAccountInfo)
+
+      _ <- setBucketAcls(bucketName, petEntityOpt.toList, List.empty)
+    } yield ()
+  }
+
+  private def setBucketAcls(bucketName: GcsBucketName, readers: List[GcsEntity], owners: List[GcsEntity]): Future[Unit] = {
+    def setBucketAndDefaultAcls(entity: GcsEntity, role: GcsRole) = {
+      for {
+        _ <- googleStorageDAO.setBucketAccessControl(bucketName, entity, role)
+        _ <- googleStorageDAO.setDefaultObjectAccessControl(bucketName, entity, role)
+      } yield ()
+    }
+
+    def flatMapList(entities: List[GcsEntity], role: GcsRole): Future[Unit] = {
```

Comment:
Can't `Future.traverse` because that runs all the Futures in parallel, and Google doesn't like when you make concurrent ACL requests on the same bucket. This works because it executes everything 1 at a time using flatMap. Not sure if there is a more elegant way of doing it..